---
layout: single
title: "[JS] 프로미스(1)"
tag: JS. Javascript, TIL
---

# Promise

 자바스크립트는 비동기 처리를 위한 하나의 패턴으로 콜백 함수를 사용한다. 하지만 전통적인 콜백 패턴은 콜백 헬로 인해 가독성이 나쁘고 비동기 처리 중 발생한 에러의 처리가 곤란하며 여러 개의 비동기 처리를 한 번에 처리하는 데도 한계가 있다.



### 비동기 처리를 위한 콜백 패턴의 단점

### 1. 콜백 지옥

비동기 함수는 함수 내부에 비동기로 동작하는 코드를 포함한 함수를 말한다. 다음 예제를 보자

```js
// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // 서버의 응답을 콘솔에 출력한다.
      console.log(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
/*
{
  "userId": 1,
  "id": 1,
  "title": "sunt aut facere ...",
  "body": "quia et suscipit ..."
}
*/
```

get함수는 서버의 응답 결과를 콘솔에 출력하기위해 작성했다. 이 함수는 어떻게 작동할까.

get함수를 호출하면 GET 요청을 전송하고 onload 이벤트 핸들러를 등록한 다음 undefined를 반환하고 즉시 종료된다. get 함수 내부의 onload 이벤드 팬들러는 get 함수가 종료된 이후에 실행된다. onload 이벤트 핸들러가 비동기로 동작하기 때문이다. 때문에 get 함수는 비동기 함수이다. get 함수가 제대로 동작하기 위해 수정해보자.

```js
// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // ① 서버의 응답을 반환한다.
      return JSON.parse(xhr.response);
    }
    console.error(`${xhr.status} ${xhr.statusText}`);
  };
};

// ② id가 1인 post를 취득
const response = get('https://jsonplaceholder.typicode.com/posts/1');
console.log(response); // undefined
```

get함수가 호출되면 XMLHttpRequest 객체를 생성하고 HTTP 요청을 초기화 한 후, HTTP 요청을 전송한다. 그리고 xhr.onload 이벤트 핸들러 프로퍼티에 이벤트 핸들러를 바인딩하고 종료한다. 이때 get 함수에 명시적인 반환문이 없으므로 get 함수는 undefined를 반환한다. xhr.onload 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러의 반환문은 get 함수의 반환문이 아니다. 

①에서 서버의 응답을 상위 스코프의 변수에 할당해도 문제는 마찬가지이다.

```js
let todos;

// GET 요청을 위한 비동기 함수
const get = url => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      // ① 서버의 응답을 상위 스코프의 변수에 할당한다.
      todos = JSON.parse(xhr.response);
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

// id가 1인 post를 취득
get('https://jsonplaceholder.typicode.com/posts/1');
console.log(todos); // ② undefined
```

 다시 한 번 get 함수를 살펴보자.<br>

비동기 함수 get이 호출되면 함수 코드를 평가하는 과정에서 get 함수의 실행 컨텍스트가 생성되고 실생 컨텍스트 스택(콜 스택)에 푸시된다. 이후 함수 코드 실행 과정이서 xhr.onload 이벤트 핸들러 프로퍼티에 이벤트 핸들러가 바인딩된다.<br>

get함수가 종료하면 get함수의 실행 컨텍스트가 콜 스택에서 팝되고, 곧바로 ②의 console.log가 호출된다. 이때 console.log의 실행 컨텍스트가 생성되어 실행 컨텍스트 스택에 푸시된다. 만약 console.log가 호춝되기 직전에 load이벤트가 발생해도 xhr.onload 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러는 절대 console.log보다 먼저 실행되지 않는다. <br>

이처럼 비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수도 없다. 따라서 비동기 함수의 처리 결과에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다. 이때 비동기 함수를 범용적으로 사용하기 위해 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백함수를 전달하는 것이 일반적이다. 필요에 따라 비동기 처리가 성공하면 호출될 콜백 함수와 비동기 처리가 실패하면 호출될 콜백 함수를 전달할 수 있다.<br>

만약 콜백 함수를 통해 비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 비동기 처리 결과를 가지고 또다시 비동기 함수를 호출해야 한다면 콜백 함수 호출이 중첩되어 복잡도가 높아지는 현상이 발생하는데, 이를 **콜백 지옥(callback hell)**이라 한다. 다음은 콜백 지옥이 발생하는 전형적인 사례다.

```js
get("/step1", (a) => {
  get(`/step2${a}`, (b) => {
    get(`/step3${b}`, (c) => {
      get(`/step4${c}`, (d) => {
        console.log(d);
      });
    });
  });
});
```

### <br><br>2. 에러 처리의 한계

비동기 처리를 위한 콜백 패턴의 문제점중에 가장 심각한 것은 에러 처리가 곤란하다는 것이다.

```js
try {
  setTimeout(() => { throw new Error('Error!'); }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다
  console.error('캐치한 에러', e);
}
```

try 코드 블록 내에서 호출한 setTimeout 함수는 1초 후에 콜백 함수가 실행되도록 타이머를 설정하고, 이후 콜백함수는 에러를 발생시킨다. 하지만 이 에러는 catch 코드 블록에서 캐치되지 않는다. <br>

이벤트 루프를 기억해보자. 비동기 함수인 setTimeout 함수의 콜백 함수가 실행될 때 setTimeout함수는 이미 콜 스택에서 제거된 상태다. 이것은 setTimeout함수의 콜백 함수를 호출한 것이 setTimeout함수가 아니라는 것을 의미한다. <br>

에러는 호출자 방향으로 전파된다. 즉 콜 스택의 아래방향(실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다. 하지만 setTimeout함수의 콜백 함수를 호출한 것은 setTimeout가 아니기 때문에 setTimeout 함수의 콜백 함수가 발생시킨 에러는 catch 블록에서 캐치되지 않는다. <br>

앞선 두 가지 문제를 극복하기 위해 프로미스가 도입되었다. 



<br><br>

# 프로미스의 생성

Promise 생성자 함수를 new 연산자와 함께 호출하면 프로미스(Promise객체)를 생성한다. Promise 생성자 함수는 비동기 처리를 수행할 콜백 함수를 인수로 전달받는데 이 콜백 함수는 resolve와 reject 함수를 인수로 전달받는다.

```js
const promise = new Promise((resolve, reject) => {
  if (success) {
    resolve("result");
  } else {
    reject("failure reason");
  }
});
```

Promise 생성자 함수가 인수로 전달받은 콜백 함수 내부에서 비동기 처리를 수행한다. 이때 비동기 처리가 성공하면 콜백 함수의 인수로 전달받은 resolve함수를 호출하고, 실패하면 reject 함수를 호출한다.  앞에서 살펴본 get함수를 프로미스를 사용해 다시 작성해보자.

```js
// GET 요청을 위한 비동기 함수
const promiseGet = url => {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open('GET', url);
    xhr.send();

    xhr.onload = () => {
      if (xhr.status === 200) {
        // 성공적으로 응답을 전달받으면 resolve 함수를 호출한다.
        resolve(JSON.parse(xhr.response));
      } else {
        // 에러 처리를 위해 reject 함수를 호출한다.
        reject(new Error(xhr.status));
      }
    };
  });
};

// promiseGet 함수는 프로미스를 반환한다.
promiseGet('https://jsonplaceholder.typicode.com/posts/1');
```

비동기함수 promiseGet은 함수 내부에서 프로미스를 생성하고 반환한다. 

<br>

프로 미스는 다음과 같이 현재 비동기 처리가 어떻게 진행되고 있는지를 나타내는 **상태**(Status) 정보를 갖는다.

| 프로미스의 상태 정보 | 의미                                  | 상태 변경 조건                   |
| -------------------- | ------------------------------------- | -------------------------------- |
| pending              | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfuled             | 비동기 처리가 수행된 상태(성공)       | `resolve` 함수 호출              |
| rejected             | 비동기 처리가 수행된 상태(실패)       | `reject` 함수 호출               |

생성된 직후의 프로미스는 기본적으로 pending 상태다. 이후 비동기 처리가 수행되면 비동기 처리 결과에 따라 다음과 같이 프로미스의 상태가 변경된다.

- 비동기 처리 성공 : resolve 함수를 호출해 프로미스를 fulfilled 상태로 변경한다.
- 비동기 처리 실패 : reject 함수를 호출해 프로미스를 reject 상태로 변경한다. 

![프로미스의 상태](../assets/images/promise.png)

정리하면 **프로미스는 비동기 처리 상태와 처리 결과를 관리하는 객체다.**

<br>

### 후속 처리 메서드

프로미스는 후속처리를 위해 메서드 then, catch, finally를 제공한다.

<br>

### <br>1. Promise.prototype.then

then 메서드는 두 개의 콜백 함수를 인수로 전달받는다.

```js
// fulfilled
new Promise(resolve => resolve('fulfilled'))
  .then(v => console.log(v), e => console.error(e)); // fulfilled

// rejected
new Promise((_, reject) => reject(new Error('rejected')))
  .then(v => console.log(v), e => console.error(e)); // Error: rejected
```

- 첫 번째 콜백 함수는 프로미스가 fulfilled 상태가 되면 호출된다. 이때 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달 받는다.
- 두 번째 콜백 함수는 프로미스가 rejected 상태가 되면 호출한다. 이때 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달 받는다.

**즉 첫 번째 콜백 함수는 비동기 처리가 성공 했을때, 두 번째 콜백 함수는 비동기 처리가 실패했을 때 호출되는 콜백 함수다.**

then 메서드의 반환값은 언제나 프로미스 객체이다.

<br>

### 2. Promise.prototype.catch

catch메서드는 한 개의 콜백 함수를 인수로 전달받아 프로미스가 rejected 상태인 경우만 호출된다.

```js
// rejected
new Promise((_, reject) => reject(new Error('rejected')))
  .catch(e => console.log(e)); // Error: rejected
```

<br>

### 3. Promise.prototype.finally

finally메서드는 한 개의 콜백 함수를 인수로 전달받아 프로미스의 성공 또는 실패와 상관없이 무조건 한 번 호출된다. **프로미스의 상태와 상관없이 공통적으로 수행해야 할 처리 내용이 있을 때 유용하다.** 마찬가지로 언제나 프로미스를 반환한다.  

```js
new Promise(() => {})
  .finally(() => console.log('finally')); // finally
```

프로미스로 구현한 비동기함수 get을 사용해 후속처리를 구현해보자

```js
const promiseGet = url => {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open('GET', url);
    xhr.send();

    xhr.onload = () => {
      if (xhr.status === 200) {
        // 성공적으로 응답을 전달받으면 resolve 함수를 호출한다.
        resolve(JSON.parse(xhr.response));
      } else {
        // 에러 처리를 위해 reject 함수를 호출한다.
        reject(new Error(xhr.status));
      }
    };
  });
};

// promiseGet 함수는 프로미스를 반환한다.
promiseGet('https://jsonplaceholder.typicode.com/posts/1')
  .then(res => console.log(res))
  .catch(err => console.error(err))
  .finally(() => console.log('Bye!'));
```

